src/LevelZeroIOGroup.cpp

```diff
@@ -0,0 +1,906 @@
+        }
+        else {
+    #ifdef GEOPM_DEBUG
+            throw Exception("LevelZeroIOGroup::" + std::string(__func__) +
+                            ": Handling not defined for " + signal_name,
+                            GEOPM_ERROR_LOGIC, __FILE__, __LINE__);
+    #endif
+        }
+        return result;
+    }
+
+    // Write to the control immediately, bypassing write_batch()
+    void LevelZeroIOGroup::write_control(const std::string &control_name,
+                                         int domain_type, int domain_idx,
+                                         double setting)
+    {
+        if (!is_valid_control(control_name)) {
+            throw Exception("LevelZeroIOGroup::" + std::string(__func__) + ": "
+                            + control_name + " not valid for LevelZeroIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_type != control_domain_type(control_name)) {
+            throw Exception("LevelZeroIOGroup::" + std::string(__func__) + ": "
+                            + control_name + ": domain_type must be " +
+                            std::to_string(control_domain_type(control_name)),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_idx < 0 ||
+            domain_idx >= m_platform_topo.num_domain(
+                                          control_domain_type(control_name))) {
+            throw Exception("LevelZeroIOGroup::" + std::string(__func__) +
+                            ": domain_idx out of range.", GEOPM_ERROR_INVALID,
+                            __FILE__, __LINE__);
+        }
+
+        if (control_name == "LEVELZERO::FREQUENCY_GPU_CONTROL" ||
+            control_name == "FREQUENCY_ACCELERATOR_CONTROL") {
+            m_levelzero_device_pool.frequency_control(domain_type, domain_idx,
+                                                      geopm::LevelZero::M_DOMAIN_COMPUTE,
+                                                      setting / 1e6);
+        }
+        else {
+    #ifdef GEOPM_DEBUG
+                throw Exception("LevelZeroIOGroup::" + std::string(__func__) +
+                                "Handling not defined for " + control_name,
+                                GEOPM_ERROR_LOGIC, __FILE__, __LINE__);
+    #endif
+        }
+    }
+
+    // Implemented to allow an IOGroup to save platform settings before starting
+    // to adjust them
+    void LevelZeroIOGroup::save_control(void)
+    {
+        for (int domain_idx = 0;
+             domain_idx < m_platform_topo.num_domain(GEOPM_DOMAIN_BOARD_ACCELERATOR);
+             ++domain_idx) {
+            // TODO: Read ALL LEVELZERO Power Limits, frequency settings, etc
+        }
+    }
+
+    // Implemented to allow an IOGroup to restore previously saved
+    // platform settings
+    void LevelZeroIOGroup::restore_control(void)
+    {
+        /// @todo: Usage of the LEVELZERO API for setting frequency, power, etc
```

Comment:
Now that we have a LevelZero abstraction for interfacing with the library, unit testing should not be a problem.